src/util/psbt/map/global.rs

```diff
@@ -0,0 +1,158 @@
+use std::collections::HashMap;
+use std::io::Cursor;
+
+use blockdata::transaction::Transaction;
+use consensus::encode::{self, Decodable, Decoder};
+use util::psbt::map::Map;
+use util::psbt::raw;
+use util::psbt::serialize::Serialize;
+use util::psbt;
+use util::psbt::Error;
+
+/// A key-value map for global data.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Global {
+    /// The unsigned transaction, scriptSigs and witnesses for each input must be
+    /// empty.
+    pub unsigned_tx: Transaction,
+    /// Unknown global key-value pairs.
+    pub unknown: HashMap<raw::Key, Vec<u8>>,
+}
+
+impl Global {
+    /// Create a Global from an unsigned transaction, error if not unsigned
+    pub fn from_unsigned_tx(tx: Transaction) -> Result<Self, psbt::Error> {
+        for txin in &tx.input {
+            if !txin.script_sig.is_empty() {
+                return Err(Error::UnsignedTxHasScriptSigs);
+            }
+
+            if !txin.witness.is_empty() {
+                return Err(Error::UnsignedTxHasScriptWitnesses);
+            }
+        }
+
+        Ok(Global {
+            unsigned_tx: tx,
+            unknown: Default::default(),
+        })
+    }
+}
+
+impl Map for Global {
+    fn insert_pair(&mut self, pair: raw::Pair) -> Result<(), encode::Error> {
+        let raw::Pair {
+            key: raw_key,
+            value: raw_value,
+        } = pair;
+
+        match raw_key.type_value {
+            0u8 => {
+                return Err(Error::DuplicateKey(raw_key).into());
+            }
+            _ => {
+                if self.unknown.contains_key(&raw_key) {
+                    return Err(Error::DuplicateKey(raw_key).into());
+                } else {
+                    self.unknown.insert(raw_key, raw_value);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn get_pairs(&self) -> Result<Vec<raw::Pair>, encode::Error> {
+        let mut rv: Vec<raw::Pair> = Default::default();
+
+        rv.push(raw::Pair {
+            key: raw::Key {
+                type_value: 0u8,
+                key: vec![],
+            },
+            value: self.unsigned_tx.serialize(),
```

Comment:
This needs to be replaced with manual serialization code to ensure 0-input transactions are serialized without witnesses. (And the corresponding unit test needs to be changed.)

```-            value: self.unsigned_tx.serialize(),
+            value: {
+                let mut ret = vec![];
+                self.unsigned_tx.version.consensus_encode(&mut ret).unwrap();
+                self.unsigned_tx.input.consensus_encode(&mut ret).unwrap();
+                self.unsigned_tx.output.consensus_encode(&mut ret).unwrap();
+                self.unsigned_tx.lock_time.consensus_encode(&mut ret).unwrap();
+                ret
+            },
         });
```
works.

Reply:
Same for the `non_witneess_utxo` in Input? Would it make sense if I just made sure that `psbt::Serialize` for `Transaction` always manually serializes?

Reply:
Hmm, `non_witness_utxo` doesn't necessarily mean that the previous tx doesn't have witnesses, right? Not sure if they are entirely irrelevant. They are for txid generation, which I think this field is primarily used for, but I wouldn't dare claim they are entirely unused.

Reply:
@dongcarl the test vectors have witnesses in `non_witness_utxo`. You need to support segwit witnesses. (And this is fine, everything is unambiguous since the referenced transactions need to be valid, i.e. have a nonzero number of outputs).

Reply:
@achow101 out of curiosity, was this deliberate?

Reply:
Yes